runtime/near-vm-logic/src/logic.rs

```diff
@@ -361,52 +352,45 @@ impl<'a> VMLogic<'a> {
     ///
     /// For nul-terminated string:
     /// `read_memory_base * num_bytes / 2 + read_memory_byte * num_bytes + utf16_decoding_base + utf16_decoding_byte * num_bytes`
-    fn get_utf16_string(&mut self, len: u64, ptr: u64) -> Result<String> {
+    fn get_utf16_string(&mut self, mut len: u64, ptr: u64) -> Result<String> {
         self.gas_counter.pay_base(utf16_decoding_base)?;
-        let mut u16_buffer;
         let max_len =
             self.config.limit_config.max_total_log_length.saturating_sub(self.total_log_length);
-        if len != u64::MAX {
-            let input = self.memory.view(&mut self.gas_counter, MemSlice { ptr, len })?;
-            if len % 2 != 0 {
-                return Err(HostError::BadUTF16.into());
+
+        let mem_view = if len == u64::MAX {
+            len = self.get_nul_terminated_utf16_len(ptr, max_len)?;
+            self.memory.view_for_free(MemSlice { ptr, len })
```

Comment:
I think for the current wasmer this is indeed an optimization. But is it really an optimization if we treat the VM as blackbox?
Bascially, I wonder if for some VM memory implementations, reading guest memory twice is more costly than reading it only once and allocating a buffer on the host.

Probably it is not worth bothering over this too much, just something that came to my mind and maybe someone else can explain it to me.

Reply:
Yes, that’s true.  If we treat VM as a black box than it can be slower.  In fact, it may be the case with WasmerMemory which doesn’t support viewing memory (i.e. view copies the memory into a vector) so in that case we first read data two bytes at a time and then read the whole thing again into a vector (on the flip side, since we know the length there are no reallocations).

I sort of convinced myself that this is a good thing to do because we care only about Wasmer2Memory.